Assertion failed in calcDir, ml/src/tree.cpp line 1190

I am training a random forest classifier with `RTrees` in the `ml` module. I have some training data that results in an assertion failure. I've distilled it down to a self-contained test case, below.


----

### Please state the information for your system

OpenCV 3.10
Linux (Ubuntu 14.04)
GCC 4.8.4, custom build

### In which part of the OpenCV library you got the issue?

opencv_ml / RTrees training

### Expected behaviour

Training completes.

### Actual behaviour

Training aborts from Python with this error:

```
OpenCV Error: Assertion failed ((int)_sleft.size() < n && (int)_sright.size() < n) in calcDir, file /home/evan/knock/opencv/modules/ml/src/tree.cpp, line 1190
Traceback (most recent call last):
  File "./train_cv.py", line 74, in <module>
    model.train(traindata)
cv2.error: /home/evan/knock/opencv/modules/ml/src/tree.cpp:1190: error: (-215) (int)_sleft.size() < n && (int)_sright.size() < n in function calcDir
```

### Additional description

### Code example to reproduce the issue / Steps to reproduce the issue


```python
import cv2
import numpy as np

samples = np.array([
    [4.13313675],
    [4.13313627],
    [4.13313627],
], dtype=np.float32);

responses = np.array([
    [2],
    [2],
    [1],
], dtype=np.int32);

varIdx = np.ones(samples.shape[1], dtype=np.uint8)
sampleIdx = np.ones(samples.shape[0], dtype=np.uint8)
varTypes = np.array([cv2.ml.VAR_NUMERICAL] * samples.shape[1] + [cv2.ml.VAR_CATEGORICAL], dtype=np.uint8)

traindata = cv2.ml.TrainData_create(
    samples=samples,
    layout=cv2.ml.ROW_SAMPLE,
    responses=responses,
    varIdx=varIdx,
    sampleIdx=sampleIdx,
    varType=varTypes)

model = cv2.ml.RTrees_create()
model.setMaxDepth(30)
model.setMinSampleCount(2)
model.setTermCriteria((cv2.TERM_CRITERIA_MAX_ITER, 40, 0.0))
model.setCalculateVarImportance(True)

model.train(traindata)
```

--- a/modules/ml/src/tree.cpp
+++ b/modules/ml/src/tree.cpp
@@ -638,7 +638,6 @@ void DTreesImpl::calcValue( int nidx, const vector<int>& _sidx )
 
 DTreesImpl::WSplit DTreesImpl::findSplitOrdClass( int vi, const vector<int>& _sidx, double initQuality )
 {
-    const double epsilon = FLT_EPSILON*2;
     int n = (int)_sidx.size();
     int m = (int)classLabels.size();
 
@@ -688,7 +687,8 @@ DTreesImpl::WSplit DTreesImpl::findSplitOrdClass( int vi, const vector<int>& _si
         rsum2 -= 2*rv*wval - w2;
         lcw[idx] = lv + wval; rcw[idx] = rv - wval;
 
-        if( values[curr] + epsilon < values[next] )
+        float value_between = (values[next] + values[curr]) * 0.5f;
+        if( value_between > values[curr] && value_between < values[next] )
         {
             double val = (lsum2*R + rsum2*L)/(L*R);
             if( best_val < val )
@@ -985,7 +985,6 @@ DTreesImpl::WSplit DTreesImpl::findSplitCatClass( int vi, const vector<int>& _si
 
 DTreesImpl::WSplit DTreesImpl::findSplitOrdReg( int vi, const vector<int>& _sidx, double initQuality )
 {
-    const float epsilon = FLT_EPSILON*2;
     const double* weights = &w->sample_weights[0];
     int n = (int)_sidx.size();
 
@@ -1021,7 +1020,8 @@ DTreesImpl::WSplit DTreesImpl::findSplitOrdReg( int vi, const vector<int>& _sidx
         L += wval; R -= wval;
         lsum += t; rsum -= t;
 
-        if( values[curr] + epsilon < values[next] )
+        float value_between = (values[next] + values[curr]) * 0.5f;
+        if( value_between > values[curr] && value_between < values[next] )
         {
             double val = (lsum*lsum*R + rsum*rsum*L)/(L*R);
             if( best_val < val )

